src/libraries/System.Web.HttpUtility/src/System/Web/HttpUtility.cs

```diff
@@ -193,10 +193,33 @@ value switch
         public static byte[]? UrlDecodeToBytes(string? str) => UrlDecodeToBytes(str, Encoding.UTF8);
 
         [return: NotNullIfNotNull(nameof(str))]
-        public static byte[]? UrlDecodeToBytes(string? str, Encoding e) => str == null ? null : UrlDecodeToBytes(e.GetBytes(str));
+        public static byte[]? UrlDecodeToBytes(string? str, Encoding e)
+        {
+            if (str == null)
+            {
+                return null;
+            }
+
+            //do not execute Encoding.GetMaxByteCount if input is too big
+            if (str.Length < 256)
+            {
+                int maxBytesCount = e.GetMaxByteCount(str.Length);
+                if (maxBytesCount < 512)
+                {
+                    //do not allocate byte[] when getting bytes from Encoding if less than stackalloc limit
+                    Span<byte> bytes = stackalloc byte[512];
+                    int encodedBytes = e.GetBytes(str, bytes);
+
+                    return UrlDecodeToBytes(bytes, 0, encodedBytes);
+                }
+            }
+
+            //no optimization possible, let Encoding allocate byte[]
```

Comment:
```suggestion
```
They're possible (e.g. renting a buffer from `ArrayPool`), we just didn't care about it enough to do it.

Reply:
I can adapt later all byte[] allocations to be rented in a separate PR if you say that it is not a problem.